File: lab10/Bank Account Command Pattern/BankApp/src/App/bank.java
package App;
import Models.Account;
import Models.Transaction;

import java.util.*;
import java.util.HashMap;
import java.util.Map;

public class bank {
    private Map<String, Account> accounts = new HashMap<>();
     public void createAccount(String accountNumber, double initialDeposite) {
         if(!accounts.containsKey(accountNumber)) {
             accounts.put(accountNumber,new Account(accountNumber,initialDeposite));
             System.out.println("Account " + accountNumber + " created with balance: " + initialDeposite);
         }
         else{
             System.out.println("Account " + accountNumber + " already exists");
         }
     }

     public void deposit(String accountNumber, double amount) {
         if (accounts.containsKey(accountNumber)) {
             accounts.get(accountNumber).deposit(amount);
         }
         else{
             System.out.println("Account " + accountNumber + " does not exist");
         }
     }

     public void withdraw(String accountNumber, double amount) {
         if(accounts.containsKey(accountNumber)){
             accounts.get(accountNumber).withdraw(amount);
         }
         else{
             System.out.println("Account " + accountNumber + " does not exist");
         }
     }

     public void showBalance(String accountNumber) {
         if(accounts.containsKey(accountNumber)){
             System.out.println("Balance for account " + accountNumber + ": " + accounts.get(accountNumber).getBalance());
         }
         else {
             System.out.println("Account " + accountNumber + " does not exist");
         }
     }
    public void showTransactionHistory(String accountNumber) {
         if(accounts.containsKey(accountNumber)){
            System.out.println("Transaction history for account " + accountNumber + ": ");
            for(Transaction transaction : accounts.get(accountNumber).getTransactions()) {
                System.out.println(transaction.getDate() + ": " + transaction.getType() + " - " + transaction.getAmount());
            }
         }
    }

}

File: Lab_10/BankApp/App/Bank.cs
using BankApp.Models;

namespace BankApp.App;

public class Bank
{
    private Dictionary<string, Account> _accounts = new();
    
    public void CreateAccount(string accountNumber, decimal initialDeposit)
    {
        if (!_accounts.ContainsKey(accountNumber))
        {
            _accounts[accountNumber] = new Account(accountNumber, initialDeposit);
            Console.WriteLine($"Account {accountNumber} created with balance: {initialDeposit}");
        }
        else
        {
            Console.WriteLine("Account already exists!");
        }
    }
    
    public void Deposit(string accountNumber, decimal amount)
    {
        if (_accounts.ContainsKey(accountNumber))
        {
            _accounts[accountNumber].Deposit(amount);
        }
        else
        {
            Console.WriteLine("Account not found!");
        }
    }

    public void Withdraw(string accountNumber, decimal amount)
    {
        if (_accounts.ContainsKey(accountNumber))
        {
            _accounts[accountNumber].Withdraw(amount);
        }
        else
        {
            Console.WriteLine("Account not found!");
        }
    }

    public void ShowBalance(string accountNumber)
    {
        if (_accounts.ContainsKey(accountNumber))
        {
            Console.WriteLine($"Balance for account {accountNumber}: {_accounts[accountNumber].Balance}");
        }
        else
        {
            Console.WriteLine("Account not found!");
        }
    }

    public void ShowTransactionHistory(string accountNumber)
    {
        if (_accounts.ContainsKey(accountNumber))
        {
            Console.WriteLine($"\nTransaction history for {accountNumber}:");
            foreach (var transaction in _accounts[accountNumber].Transactions)
            {
                Console.WriteLine($"{transaction.Date}: {transaction.Type} - {transaction.Amount}");
            }
        }
        else
        {
            Console.WriteLine("Account not found!");
        }
    }
}